File: Residential_Controller.js
//"*************************************************************"
//"                 -  TEST SCENARIOS -                         "
//"*************************************************************"

function testBattery() {

    var testScenario = initSystem(10, 2);
    var elevatorOne = testScenario.column.elevatorsCulumn[0];
    var elevatorTwo = testScenario.column.elevatorsCulumn[1];


    //Scenario 1 

    // elevatorOne.currentPosition = 2;                            // elevator 1 Idle at floor 2
    // elevatorOne.state = "Idle";
    // elevatorOne.elevatorDirection = "Up";

    // elevatorTwo.currentPosition = 6;                            //  elevator 2  Idle at floor 6
    // elevatorTwo.state = "Idle";
    // elevatorTwo.elevatorDirection = "Down";

    // var elevator = testScenario.RequestElevator(3, "Up");         //someone is on floor 3 and requests the 7th floor
    // testScenario.RequestFloor(elevator, 7);                       // and requests the 7th floor

    // console.log("")

    // console.log("")
    // console.log("---Test scenario 1 : Ok++")                  // elevator 1 is expected to be sen
    // console.log("")

    // console.log("++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")

    //Scenario 2 

    // elevatorOne.currentPosition = 10;                    //elevator 1 idle at floor 10
    // elevatorOne.state = "Idle";
    // elevatorOne.elevatorDirection = "Up";

    // elevatorTwo.currentPosition = 3;                     //  elevator 2 idle at floor 3,
    // elevatorTwo.state = "Idle";
    // elevatorTwo.elevatorDirection = "Up";

    // var elevator = testScenario.RequestElevator(1, "Up");        // someone is on the 1st floor 
    // testScenario.RequestFloor(elevator, 6);                      //and requests the 6th floor = elevator 2 should be sent = ok

    // elevator = testScenario.RequestElevator(3, "Up");            // someone else is on the 3rd floor 
    // testScenario.RequestFloor(elevator, 5);                      // and requests the 5th floor.Elevator 2 should be sent.

    // elevator = testScenario.RequestElevator(9, "Down");             //on floor 9
    // testScenario.RequestFloor(elevator, 2);                       // and wants to go down to the 2nd floor= Elevator 1 should be sent.

    // console.log("");
    // console.log("---Test scenario 2 : Ok++");
    // console.log("");
    // console.log("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")

    //Scenario 3 

    elevatorOne.currentPosition = 10; //elevator 1 idle at floor 10 
    elevatorOne.state = "Idle";
    elevatorOne.elevatorDirection = "Down";

    elevatorTwo.currentPosition = 3; //elevator 2 moving from floor 3 to floor 6
    elevatorTwo.state = "Moving"; // and wants to go down to the 3rd floor. Elevator 2 should be sent => Ok
    elevatorTwo.elevatorDirection = "Up";

    // someone is on the 1st floor 
    var elevator = testScenario.RequestElevator(3, "Down"); //and requests the 6th floor = elevator 2 should be sent = ok

    testScenario.RequestFloor(elevator, 2);             //on floor 9
    elevator = testScenario.RequestElevator(10, "Down"); // and wants to go down to the 2nd floor= Elevator 1 should be sent.=ok
    testScenario.RequestFloor(elevator, 3);                // someone else is on the 10th floor and wants to go down  to the 3rd 


    console.log("");
    console.log("---Test scenario 3 : Ok++")
    console.log("");
    console.log("+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++")

}

//*************************************************************"
//             - ELEVATORS SYSTEM INITIALISATION  -            "
//**************************************************************"

function initSystem(nbFloors, nbElevators) {
    var syst = new ElevatorController(nbFloors, nbElevators);

    return syst;
}
class Column {
    constructor(nbFloors, nbElevators) {

        this.nbFloors = nbFloors;
        this.nbElevators = nbElevators;
        this.elevatorsCulumn = [];

        for (let i = 0; i < this.nbElevators; i++) {

            let elevator = new Elevator(i + 1, "Idle", 1, "Up");
            this.elevatorsCulumn.push(elevator);
        }
    }
}

class Elevator {

    constructor(elevatorId, state, currentPosition, elevatorDirection) {

        this.elevatorId = elevatorId;
        this.state = state;
        this.currentPosition = currentPosition;
        this.elevatorDirection = elevatorDirection;
        this.elevatorFloorsList = [];

    }

    // "*************************************************************"
    // "                 -  ELEVATORS STOPS LIST -                  "
    // "*************************************************************"

    // Add the request to the elevators list

    addToList(RequestedFloor) {

        this.elevatorFloorsList.push(RequestedFloor);
        this.sortList();
        this.operate_elevator(RequestedFloor);
    }

    // Sort the list depending on the direction of the elevator

    sortList() {
        if (this.elevatorDirection === "Up") {
            this.elevatorFloorsList.sort();
        } else if (this.elevatorDirection === "Down") {
            this.elevatorFloorsList.sort();
            this.elevatorFloorsList.reverse();
        }
        return this.elevatorFloorsList;
    }

    // Move the elvators to their destination

    operate_elevator(RequestedFloor) {
        while (this.elevatorFloorsList > 0) {
            if (RequestedFloor === this.currentPosition) {
                this.openDoors();
                this.state = "Moving";

                this.elevatorFloorsList.shift();
            } else if (RequestedFloor < this.currentPosition) {
                this.state = "Moving";
                console.log("Elevator " + this.elevatorId + " is selected");
                console.log("Elevator " + this.elevatorId, this.state);
                console.log("");
                this.Direction = "Down";
                this.moveDown(RequestedFloor);
                this.state = "Stopped";
                console.log("");
                console.log("Elevator " + this.elevatorId, this.state);
                console.log("");
                this.openDoors();
                this.elevatorFloorsList.shift();
            } else if (RequestedFloor > this.currentPosition) {
                delay(1000);
                this.state = "Moving";
                console.log("");
                console.log("Elevator " + this.elevatorId + " is selected");
                console.log("Elevator " + this.elevatorId, this.state);
                console.log("");
                this.Direction = "Up";
                this.moveUp(RequestedFloor);
                this.state = "Stopped";
                console.log("");
                console.log("Elevator " + this.elevatorId, this.state);
                console.log("");

                this.openDoors();

                this.elevatorFloorsList.shift();
            }
        }
        if (this.elevatorFloorsList === 0) {
            this.state = "Idle";
        }
    }

    // "*************************************************************"
    // "               -  Floor Request & Call Buttons Control -     "
    // "*************************************************************"
    Request_floor_button(RequestedFloor) {

        this.RequestedFloor = RequestedFloor;
        this.floorLights = floorLights;
    }
    Call_floor_button(floor, Direction) {

        this.floor = floor;
        this.Direction = Direction;
    }

    // "*************************************************************"
    // "                 -  ELEVATORS DOORS CONTROL -                "
    // "**************************************************************"

    openDoors() {
        delay(1000);
        console.log("Doors opening");
        console.log("Doors opened");
        console.log("");
        console.log("Button light-off");
        delay(1000);

        console.log("");
        delay(1000);
        this.Close_door();
    }
    Close_door() {
        console.log("Closing doors");
        console.log("Doors' not obstructed : OK")
        console.log("Elevator's capacity check: OK ")
        console.log("Doors Closed")
        delay(1000);
    }

    // "*************************************************************"
    // "                 -  ElEVATORS MOVEMENTS -                      "
    // "*************************************************************"

    moveUp(RequestedFloor) {
        console.log("Floor : " + this.currentPosition);
        delay(1000);
        while (this.currentPosition !== RequestedFloor) {
            this.currentPosition += 1;
            console.log("Floor : " + this.currentPosition);

            delay(1000);
        }
    }

    moveDown(RequestedFloor) {
        console.log("Floor : " + this.currentPosition);
        delay(1000);
        while (this.currentPosition !== RequestedFloor) {
            this.currentPosition -= 1;
            console.log("Floor : " + this.currentPosition);

            delay(1000);
        }
    }
}
// "*************************************************************"
// "                 -  ElEVATORS CONTROL -                      "
// "*************************************************************"

class ElevatorController {
    constructor(nbFloors, nbElevators) {
        this.nbFloors = nbFloors;
        this.nbElevators = nbElevators;
        this.column = new Column(nbFloors, nbElevators);


        console.log("System is starting");
    }

    // Request an a floor

    RequestElevator(floor, Direction) {
        delay(1000);
        console.log("");
        console.log("Request elevator to floor : ", floor);
        delay(1000);
        console.log("");
        console.log("Call button illuminated");
        delay(1000);

        let elevator = this.selectElevator(floor, Direction);
        elevator.addToList(floor);
        return elevator;
    }

    // Call an elevator 

    RequestFloor(elevator, RequestedFloor) {
        delay(1000);
        console.log("");
        console.log("The floor number " + RequestedFloor + " is requested");
        delay(1000);
        console.log("");
        console.log("Request button illuminated");
        delay(1000);
        elevator.addToList(RequestedFloor);

    }

    // "*************************************************************"
    // "                 -  SELECT AN ELEVATOR -                 "
    // "**************************************************************"

    selectElevator(floor, Direction) {

        let suitableElevator = null;
        let shortestDistance = 1000;
        for (let i = 0; i < this.column.elevatorsCulumn.length; i++) {
            let elevator = this.column.elevatorsCulumn[i];

            if (floor === elevator.currentPosition && (elevator.state === "Stopped" || elevator.state === "Idle" || elevator.state === "Moving")) {
                return elevator;
            } else {
                let distance = Math.abs(floor - elevator.currentPosition);
                if (shortestDistance > distance) {
                    shortestDistance = distance;
                    suitableElevator = elevator;

                    if (elevator.Direction === Direction) {
                        suitableElevator = elevator;
                    }
                }
            }
        }
        return suitableElevator;
    }
}

// "*************************************************************"
// "                 -  DELAY CLOSING/OPENING DOORS -            "
// "*************************************************************"
function delay(milliseconds) {
    var start = new Date().getTime();
    for (var i = 0; i < 1e7; i++) {
        if (new Date().getTime() - start > milliseconds) {
            break;
        }
    }
}

File: Residential_Controller.py
import time


class columnController:
    def __init__(self, nbFloors, nbElevators):
        self.nbFloors = nbFloors
        self.nbElevators = nbElevators
        self.column = Column(nbFloors, nbElevators)
       


    def RequestElevator(self, floor, direction):
        time.sleep(1)
        print("")
        print("Request elevator to floor : ", floor)
        time.sleep(1)
        print("")
        print("Call button illuminated")
        time.sleep(1)
        elevator = self.selectElevator(floor, direction)
        elevator.addToList(floor)
        return elevator

    def RequestFloor(self, elevator, RequestedFloor):
        time.sleep(1)
        print("")
        print("Requested floor to ", RequestedFloor)
        time.sleep(1)
        print("")
        print("Request button illuminated")
        time.sleep(1)
        elevator.addToList(RequestedFloor)


    def selectElevator(self, floor, direction):
        bestElevator = None
        shortest_distance = 1000
        for elevator in (self.column.elevatorsColumn):
            if (floor == elevator.elevatorCurrentFloor and (elevator.elevatorStatus == "Stopped" or elevator.elevatorStatus == "Idle" or elevator.elevatorStatus == "Moving")):
                return elevator
            else:
                ref_distance = abs(floor - elevator.elevatorCurrentFloor)
                if shortest_distance > ref_distance:
                    shortest_distance = ref_distance
                    bestElevator = elevator

                elif elevator.direction == direction:
                    bestElevator = elevator

        return bestElevator


class Column:
    def __init__(self, nbFloors, nbElevators):
        self.nbFloors = nbFloors
        self.nbElevators = nbElevators
        self.elevatorsColumn = []
        for i in range(nbElevators):
            elevator = Elevator(i+1, "Idle", 1, "Up")
            self.elevatorsColumn.append(elevator)
            
class Elevator:
    def __init__(self, elevatorId, elevatorStatus, elevatorCurrentFloor, elevatorDirection):
        self.elevatorId = elevatorId
        self.elevatorStatus = elevatorStatus
        self.elevatorCurrentFloor = elevatorCurrentFloor
        self.elevatorDirection = elevatorDirection
        self.floor_list = []

    def addToList(self, RequestedFloor):
        self.floor_list.append(RequestedFloor)
        self.sort()
        self.operate_elevator(RequestedFloor)


    def sort(self):
        if self.elevatorDirection == "Up":
            self.floor_list.sort()
        elif self.elevatorDirection == "Down":
            self.floor_list.sort()
            self.floor_list.reverse()
        return self.floor_list


    def operate_elevator(self, RequestedFloor):
        while (len(self.floor_list) > 0):
            if ((RequestedFloor == self.elevatorCurrentFloor)):
                self.openDoors()
                self.elevatorStatus = "Moving"
                self.floor_list.pop()
            elif (RequestedFloor < self.elevatorCurrentFloor):

                self.elevatorStatus = "Moving"
                print("")
                print("Elevator number ", self.elevatorId,"is ", self.elevatorStatus)
                print("")
                self.direction = "Down"
                self.moveDown(RequestedFloor)
                self.elevatorStatus = "Stopped"
                print("")
                print("Elevator number ", self.elevatorId,"is ", self.elevatorStatus)
                print("")
                self.openDoors()
                self.floor_list.pop()
                
            elif (RequestedFloor > self.elevatorCurrentFloor):

                time.sleep(1)
                self.elevatorStatus = "Moving"
                print("")
                print("Elevator number ", self.elevatorId,"is ", self.elevatorStatus)
                print("")
                self.direction = "Up"
                self.moveUp(RequestedFloor)
                self.elevatorStatus = "Stopped"
                print("")
                print("Elevator number", self.elevatorId,"is ", self.elevatorStatus)
                print("")

                self.openDoors()

                self.floor_list.pop()

        if self.floor_list == 0:
            self.elevatorStatus = "Idle"



    def openDoors(self):
        time.sleep(1)
        print("Open Door")
        print("")
        print("Button deactivated")
        time.sleep(1)
        print("")
        time.sleep(1)
        self.Close_door()

    def Close_door(self):
        print("Close door")
        print("Doors not obstructed")
        print("Elevator's capacity check: OK ")
        print("Doors Closed")
        time.sleep(1)



    def moveUp(self, RequestedFloor):
        print("Floor : ", self.elevatorCurrentFloor)
        time.sleep(1)
        while(self.elevatorCurrentFloor != RequestedFloor):
            self.elevatorCurrentFloor += 1
            print("Floor : ", self.elevatorCurrentFloor)
            time.sleep(1)

    def moveDown(self, RequestedFloor):
        print("Floor : ", self.elevatorCurrentFloor)
        time.sleep(1)
        while(self.elevatorCurrentFloor != RequestedFloor):
            self.elevatorCurrentFloor -= 1
            print("Floor : ", self.elevatorCurrentFloor)

            time.sleep(1)


class Call_button:
    def __init__(self, floor, direction):
        self.floor = floor
        self.direction = direction
        self.light = False


class Floor_button:
    def __init__(self, RequestedFloor):
        self.RequestedFloor = RequestedFloor


"*************************************************************"
"                 -  TEST SCENARIOS -                         "
"*************************************************************"


# Scenario 1 

controllerOne = columnController(10, 2)

elevatorNumOne = controllerOne.column.elevatorsColumn[0]
elevatorNumTwo = controllerOne.column.elevatorsColumn[1]

elevatorNumOne.elevatorCurrentFloor = 2
elevatorNumOne.elevatorStatus = "Moving"
elevatorNumOne.elevatorDirection = "Down"

elevatorNumTwo.elevatorCurrentFloor = 6
elevatorNumTwo.elevatorStatus = "Moving"
elevatorNumTwo.elevatorDirection = "Down"

print("")
print("---Test scenario 1 starts---")
print("")
elevator = controllerOne.RequestElevator(5, "Up")
controllerOne.RequestFloor(elevator, 7)
print("")
print("---The end of the test scenario 1---")
print("")


# Scenario 2 

controllerTwo = columnController(10, 2)

elevatorNumOne = controllerTwo.column.elevatorsColumn[0]
elevatorNumTwo = controllerTwo.column.elevatorsColumn[1]

elevatorNumOne.elevatorCurrentFloor = 10
elevatorNumOne.elevatorStatus = "Moving"
elevatorNumOne.elevatorDirection = "Down"
elevatorNumTwo.elevatorCurrentFloor = 3
elevatorNumTwo.elevatorStatus = "Moving"
elevatorNumTwo.elevatorDirection = "Down"

print("")
print("---Test scenario 2 starts---")
print("")


elevator = controllerTwo.RequestElevator(1, "Up")
controllerTwo.RequestFloor(elevator, 6)
elevator = controllerTwo.RequestElevator(3, "Up")
controllerTwo.RequestFloor(elevator, 5)
elevator = controllerTwo.RequestElevator(9, "Down")
controllerTwo.RequestFloor(elevator, 2)
print("")
print("The end of the test scenario 2 ")
print("")


print("+++++++++++++++++++++++++++++++++++++++")

# Scenario 3 

controllerThree = columnController(10, 2)

elevatorNumOne = controllerThree.column.elevatorsColumn[0]
elevatorNumTwo = controllerThree.column.elevatorsColumn[1]

elevatorNumOne.elevatorCurrentFloor = 10
elevatorNumOne.elevatorStatus = "Moving"
elevatorNumOne.elevatorDirection = "Down"

elevatorNumTwo.elevatorCurrentFloor = 3
elevatorNumTwo.elevatorStatus = "Moving"
elevatorNumTwo.elevatorDirection = "Down"



print("+++++++++++++++++++++++++++++++++++++++")

print("")
print("---Test scenario 3 starts---")
print("")

elevator = controllerThree.RequestElevator(10, "Down")
controllerThree.RequestFloor(elevator, 3)

elevator = controllerThree.RequestElevator(3, "Down")
controllerThree.RequestFloor(elevator, 2)

print("")
print("The end of the test scenario 3")
print("")
